Assignments/Splitwise/Python/main.py

```diff
@@ -0,0 +1,103 @@
+from user import user
+from typing import List
+def main():
+    p = ProcessInput()
+    while True:
+        #process each line in input
+        try:
+            inputList = input().strip().split(' ')
+            if inputList[0] == 'ADD':
+                p.processAdd(inputList)
+            if inputList[0] == 'EXPENSE':
+                p.processExpense(inputList)
+            if inputList[0] == 'SHOW':
+                p.processShow(inputList)
+            if inputList[0] == '':
+                exit(0)
+        except (EOFError):
+            exit(0)
+
+
+def roundNumber(number):
+    if int(number) == number:
+        return int(number)
+    else:
+        return round(number, 2)
+    
+
+class ProcessInput:
```

Comment:
I understand you are clubbing all of the expense types into a single expense type. However, this code is not extensible. So you are not living upto the expectations as mentioned in the problem statement.

Example: If I have to add another feature like whenever, I add an expense, I want to have a choice of adding a writeup. The current code does not serve its purpose.

Please read the expectations tab of ``` README.md ``` carefully

Reply:
So would it mean that I can possibly modularise the EXPENSE as it's own class? 
Now if we can modularise the expenses, then I can probably refactor the current `ProcessInput` class.

Here's the approach.
 - `class Group` will be formed for each group. (As an extension to requirements, there can be multiple groups of people to handle).
    - `users` attribute will store all the users in the group.
    - `expenses` attribute will store all the EXPENSE class instances, this will help in tracking all the transactions. (This is not ideal for memory, probably storing each expense into a file or database would be better, and instead of tracking a bunch of expenses, each `addExpense` function call would store the expense into a file/DB. But then this applies to all the attributes in general.)
    - `balances` attribute (previously `owes`) stays the same tracking the current balances among the users. (This is probably a bit expensive in terms memory but will be faster in execution.)
    - `processAdd` method can now be the constructor of the `class Group` this will create a group of users. This also provides the flexibility of adding users to the group later.
    - `processExpense` method can now be `addExpense` with the same parameters.
    - `processShow` method can now be `showBalances` with the same parameters.
 - `class Expense` will be tracking each transaction.
    - `expenseType` attribute will store the type of expense (i.e. EQUAL/PERCENT/EXACT/SHARE). I can use an `enum` for the same.
    - `users` attribute will track all the users involved in the transaction. (Helpful for the optional passbook requirement).
    - `expenseBy` attribute will store the information on who paid for the expense.
    - `splitAmong` (dictionary) attribute will store all the users among which the expense is to be split. Key can be the user ID and its value can be the amount that the user owes.


Please provide your feedback on this and then I'll start implementing this approach.

Reply:
You are really close to the solution. Kudos!! 👍 . Also to an ice on the cake, Expense class can also be having types. You might also want to consider splitwise as a class which handles these things. If you want to have custom features of each of these expense types (example: If I come up with a requirement that any store you go, you need to have an attachment of receipt)., this can also serve it

Reply:
May be you can refactor ``` Expense ```. You are required to show a Passbook of a user. CUrrent implementation cannot still suffice

Reply:
To implement `passbook` since it must be maintained for each user. I thought of refactoring the `users` class to contain a `passbook` attribute. This `passbook` is a list that contains instances of `Expense` class